frontend/depgraph.lua

```diff
@@ -45,72 +93,92 @@ function DepGraph:removeNode(node_key)
     -- one in their override=, that have added themselves in
     -- this node's .deps). We don't want to lose these
     -- dependencies if we later re-addNode this node.
-    local node = self.nodes[node_key]
+    local node, index = self:getNode(node_key)
     if node then
         if not node.deps or #node.deps == 0 then
-            self.nodes[node_key] = nil
+            -- We need to clear *all* references to that node for it to be GC'ed...
+            node = nil --luacheck: ignore
+            table.remove(self.nodes, index)
         end
     end
-    -- But we should remove it from the .deps of other nodes.
-    for curr_node_key, curr_node in pairs(self.nodes) do
-        if curr_node.deps then
-            local remove_idx
-            for idx, dep_node_key in ipairs(self.nodes) do
+    -- But we should remove it from the .deps of *other* nodes.
+    for i, _ in ipairs(self.nodes) do
+        local curr_node = self.nodes[i]
+        -- Is not the to be removed node, and has deps
+        if curr_node.key ~= node_key and curr_node.deps then
+            -- Walk that node's deps to check if it depends on us
+            for idx, dep_node_key in ipairs(curr_node.deps) do
+                -- If it did, wipe ourselves from there
                 if dep_node_key == node_key then
-                    remove_idx = idx
+                    -- Wipe all refs
+                    curr_node.deps[idx] = nil
+                    table.remove(self.nodes[i].deps, idx)
```

Comment:
Making a hole in the array ? Why not `table.remove(curr_node.deps , idx)` ?
Oh, but that's what that 2nd line does, as curr_node = self.nodes[i].
So, that first line is not needed and confusing.

I think you got yourself too much drown in this thinking about "refs" :/ I usually don't need to think like that.

Reply:
"technically for show" ? What does that mean ? :)

What annoys me there is not so much the fact that you uneededly set it to `nil` first - but that on the first line, you access the thing via `curr_node.deps` - and on the second line via `self.nodes[i].deps` - which makes the reader think you're dealing with 2 different things - and spend some time verifying that they are indeed _not_ different things.
So, either the reader (me) is confused - or the author was :)

One of these make it less confusing:
```lua
curr_node.deps[idx] = nil -- uneeded but artistically for show :)
table.remove(curr_node.deps, idx)
```

```lua
self.nodes[i].deps[idx] = nil -- uneeded but artistically for show :)
table.remove(self.nodes[i].deps, idx)
```
Or I don't really get it ?!

Reply:
The nil assign is technically useless here because that local is never re-user in the function.

But that leads to an interesting query:

What happens to `curr_node.deps[idx]` after the `table.remove`? Is the ref still pointing at old content? Or would accessing it crash because it's attempting to index a nil? Or would it be pointing to something else because the array has been re-ordered by remove?

Reply:
On the other hand, the ref being used for the  `table.remove` doesn't matter, because it'll get at the actual table (i.e., it deref's).

Hence, using the "main" reference (self's) for it was also to make that fact more obvious (i.e., also documentation).

Reply:
Okay, it's actually a terrible idea anyway because it makes the array sparse, which is Very Bad :).

(c.f., other comments above).

Reply:
> What happens to curr_node.deps[idx] after the table.remove? 

table.remove returns it, and shift everything after it in the array. And if you're not assigning it (e.g. you're not doing `local thatThing = table.remove(curr_node.deps, idx)`), and if it's not referenced elsewhere, its ref count should be down to 0 and it'll be candidate for gc.
If you previously accessed it `local oldThing = curr_node.deps[idx]`, you've added 1 to the ref count, that thing still exists (it's _old content_ if you want, but it's that thing :) it's just no more in the table.

> As such, the nil is for documentation purposes, to make it clear that that local ref ought to be dead, 

Possibly excessive and confusing documentation to me :) once you get the Lua way of things describe above.

But thanks for the simplication. It all looks fine to me, clear, and nothing confusing :)